plugins/browser-plugin-media/src/mediaTracking.ts

```diff
@@ -0,0 +1,200 @@
+    adBreak?: MediaPlayerAdBreakUpdate
+  ): { event: SelfDescribingJson; context: SelfDescribingJson[] }[] {
+    // update state
+    this.updateMediaPlayer(eventType, mediaPlayer);
+    if (eventType !== undefined) {
+      this.adTracking.updateForThisEvent(eventType, this.mediaPlayer, ad, adBreak);
+    }
+    this.session?.update(eventType, this.mediaPlayer, this.adTracking.adBreak);
+    this.pingInterval?.update(this.mediaPlayer);
+
+    // build context entities
+    let context = [buildMediaPlayerEntity(this.mediaPlayer)];
+    if (this.session !== undefined) {
+      context.push(this.session.getContext());
+    }
+    if (this.customContext !== undefined) {
+      context = context.concat(this.customContext);
+    }
+    context = context.concat(this.adTracking.getContext());
+
+    // build event types to track
+    let eventTypesToTrack: MediaPlayerEventType[] = [];
+    if (eventType !== undefined && this.shouldTrackEvent(eventType)) {
+      eventTypesToTrack.push(eventType);
+    }
+    if (this.shouldSendPercentProgress()) {
+      eventTypesToTrack.push(MediaPlayerEventType.PercentProgress);
+    }
+
+    // update state for events after this one
+    if (eventType !== undefined) {
+      this.adTracking.updateForNextEvent(eventType);
+    }
+
+    return eventTypesToTrack.map((eventType) => {
+      return {
+        event: buildMediaPlayerEvent(eventType, this.label),
+        context: context,
+      };
+    });
+  }
+
+  private updateMediaPlayer(
+    eventType: MediaPlayerEventType | undefined,
+    mediaPlayer: MediaPlayerUpdate | undefined
+  ) {
+    if (mediaPlayer !== undefined) {
+      this.mediaPlayer = {
+        ...this.mediaPlayer,
+        ...mediaPlayer,
+      };
+    }
+    if (eventType == MediaPlayerEventType.Play) {
+      this.mediaPlayer.paused = false;
+    }
+    if (eventType == MediaPlayerEventType.Pause) {
+      this.mediaPlayer.paused = true;
+    }
+    if (eventType == MediaPlayerEventType.End) {
+      this.mediaPlayer.paused = true;
+      this.mediaPlayer.ended = true;
+    }
+    if (this.mediaPlayer.duration && this.mediaPlayer.duration > 0) {
+      this.mediaPlayer.percentProgress = Math.floor((this.mediaPlayer.currentTime / this.mediaPlayer.duration) * 100);
+    }
+  }
+
+  private shouldSendPercentProgress(): boolean {
+    if (
+      this.boundaries === undefined ||
+      this.mediaPlayer.percentProgress === undefined ||
+      this.mediaPlayer.percentProgress === null ||
+      this.mediaPlayer.paused
+    ) {
+      return false;
+    }
+
+    let boundaries = this.boundaries.filter((b) => b <= (this.mediaPlayer.percentProgress ?? 0));
+    if (boundaries.length == 0) {
+      return false;
+    }
+    let boundary = Math.max(...boundaries);
+    if (!this.sentBoundaries.includes(boundary)) {
+      this.sentBoundaries.push(boundary);
+      return true;
+    }
+    return false;
+  }
+
+  private shouldTrackEvent(eventType: MediaPlayerEventType) {
```

Comment:
This method seems to perform two fairly distinct functions:
- controlling the state of `this.isSeeking`
- checking if an `eventType` should be tracked

I see why, as we don't want to track a `SeekStart` if we are already seeking, but maybe we could either give the method a name that better captures all the functions it performs, or extract the state-setting into it's own small method to call from within `shouldTrackEvent`, something like:

```
private updateSeekState(eventType: MediaPlayerEventType): void {
    if (eventType == MediaPlayerEventType.SeekStart) {
        if (this.isSeeking) {
            return false;
        }

        this.isSeeking = true;
    } else if (eventType == MediaPlayerEventType.SeekEnd) {
        this.isSeeking = false;
    }
}

private shouldTrackEvent(eventType: MediaPlayerEventType): boolean {
    updateSeekState(eventType);

    if (this.captureEvents !== undefined && !this.isSeeking && !this.captureEvents.includes(eventType)) {
      return false;
    }
}
```